### ISSUE TITLE
[ActiveRecord] schema_cache.yaml does not account for the database schema (tested in postgres)

### ISSUE THREAD

**Issue by sbeckeriv** [rails/rails#41551]
Dearest maintainer, 

I hope all is well. I found this issue within my project and thought it best to share back. I have models under a different database schema and the schema cache yml file does not properly provided the caching for the schema table lookup. This can cause every request to hit the database for the sql again. 

### Steps to reproduce
I found it in rails 5.2 but it still has the same issue in 6.1.  I only tested postgres. I am unsure if it can be done in sqlite

My set of files from a new rails project: 

configure you database for postgres. I also have `schema_search_path: "public,admin" ` and i am unsure if that is required.
```
default: &default
  adapter: postgresql
  schema_search_path: "public,admin"
  encoding: utf8
  host: "localhost"
  pool: 5

development:
  <<: *default
  host: "localhost"
  database: schematest 
```

db/migrate/20210225175052_schmeatest.rb 
```
class Schmeatest < ActiveRecord::Migration[6.1]
  def change
    create_schema "admin"
    create_table "admin.tests" do |t|
      t.string :name, null: false
    end
  end
end
```

app/models/admin/test.rb 
```
module Admin
    def self.table_name_prefix
      'admin.'
    end
end

module Admin
  class Test < ApplicationRecord
    validates :name, presence: true
  end
end
```

note: `config.active_record.schema_format = :sql` is required with db schema if you want to inspect the dump.

Run:
`bundle exec rake db:migrate`
`bundle exec rake db:schema:cache:dump`

### Expected behavior
I would expect the yaml file to include a table that is admin.tests. 

### Actual behavior
It does not contain a table for admin.tests but only tests. 

It is hard to test/see because rails does not log these calls. (that is feature for another day)
https://github.com/rails/rails/blob/291a3d2ef29a3842d1156ada7526f4ee60dd2b59/activerecord/lib/active_record/connection_adapters/schema_cache.rb#L82
this file here if you turn on rails to use the cache should return early. 

The name passed in to here when loading `Admin::Test.last` is `admin.tests`

I think the issue is deeper then the check in the cache. Following where the schema data gets its information from you hit this sql.  https://github.com/rails/rails/blob/291a3d2ef29a3842d1156ada7526f4ee60dd2b59/activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb#L758
It appears this sql does not account for the schema setup?

My list of table names unique and I have patched the schema cache dumper to include "tests" and "admin.tests". This has appeared to solve my caching issue for the moment. 

### System configuration
**Rails version**:
Rails 5.2 and 6.1 tested

**Ruby version**:
2.6

I hope this makes sense and if nothing else it helps someone else. 
Thanks again for reviewing the issue and your work,

Becker

**Comment by p8** [comment]
@sbeckeriv Thanks for reporting the issue.
This seems related to #41575

**Comment by sbeckeriv** [comment]
Hello @p8,

It might be related but I do not know the underlying code well enough to agree. The issue linked appears to be a problem with the sql generation code. I have tracked my issue down to the database reflection code and the schema dumper. The dump for schema cache does not use any of the sql generation code.  It uses the in memory object which is feed data directly from the databases. Issue #41575  might be a problem when I am upgrading thanks for bringing it to my attention. 

Thanks
Becker

**Comment by fatkodima** [comment]
> This can cause every request to hit the database for the sql again.

Related query will run **only once** and the result will be cached in `SchemaCache`. So, yaml file with cache is basically not effective for models with explicitly specified schemas in `table_name`s.

> It appears this sql does not account for the schema setup?

Yes, it returns a flat list of table names and there is no way to account for schemas without introducing backward incompatibilities.

If you do not have duplicate table names both in `public` and `admin`, it is better to remove `self.table_name_prefix` and let the activerecord use the default table name.

**Comment by sbeckeriv** [comment]
@fatkodima Thank you for the reply. it has been a while since I looked the issue. I am not sure that the `SchemaCache` is accounting for the prefix even when it is loaded from the table list. 
https://github.com/rails/rails/blob/291a3d2ef29a3842d1156ada7526f4ee60dd2b59/activerecord/lib/active_record/connection_adapters/schema_cache.rb#L82

so this line  `return @data_sources[name] if @data_sources.key? name` @data_sources has a key "test" but the name variable is "admin.test" which is not in the list. `admin.test` will be written to the @data_source after it hits the database via data_source_exists?(name). This happens once per uncached connection per model using a prefix. I did not reconfirm this locally. 

My current project does not have overlapping table names. removing the table_name_prefix would not help in this case.

I understand It is a flat list. In my current project the table names are all prefixed in a way where I can write both `test` and `admin.test`. If it could be detected from the database this concept could provide a transition strategy to move to a consistent use of the schema prefixes. 

Thanks again.
Becker
